go-pkg/ds/permission.go

```diff
@@ -1,27 +1,94 @@
 package ds
 
 import (
+	"crypto/ecdsa"
+	"errors"
+
+	ethereum "github.com/ethereum/go-ethereum/crypto"
+	"github.com/jinfwhuang/ds-toolkit/go-pkg/encrypt"
 	protods "github.com/jinfwhuang/ds-toolkit/proto/ds"
 )
 
 // Do I have permission to the DataBlob
-func (u *User) checkPerm(data *protods.DataBlob) bool {
-	panic("not implemented")
+func checkPerm(data *protods.DataBlob, pubKey *ecdsa.PublicKey) bool {
+	pubKeyBytes := ethereum.CompressPubkey(pubKey)
+	_, err := findUserKey(data.Keys, pubKeyBytes)
+	return err == nil
 }
 
 // Get the decrypted data in EncryptedData
-func (u *User) extractData(data *protods.DataBlob) []byte {
-	panic("not implemented")
+func extractData(data *protods.DataBlob, privKey *ecdsa.PrivateKey) ([]byte, error) {
+	pubKeyBytes := ethereum.CompressPubkey(&privKey.PublicKey)
+	userKey, err := findUserKey(data.Keys, pubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+	dataKey, err := recoverHiddenDataKey(userKey, privKey.D.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedData, err := encrypt.Decrypt(dataKey, data.Iv, data.EncryptedData)
+	if err != nil {
+		return nil, errors.New("failed to decrypt data")
+	}
+
+	// AES blocks are padded, we need to get rid of the padding
+	unpaddedDecryptedData := removeBlockCipherPadding(decryptedData)
```

Comment:
Is this always correct?

How about making use of data length, which is available in DataBlog?

Reply:
I checked with 3-4 edge cases I could think about and it was fine. But you are absolutely right, I forgot about the DataLen field, much more elegant solution